src/exception/exception-manifest/util.ts

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright (c) 2024 The Ontario Institute for Cancer Research. All rights reserved
+ *
+ * This program and the accompanying materials are made available under the terms of
+ * the GNU Affero General Public License v3.0. You should have received a copy of the
+ * GNU Affero General Public License along with this program.
+ *  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT
+ * SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
+ * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+ * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
+ * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import { DeepReadonly } from 'deep-freeze';
+import { Donor, FollowUp, Specimen, Treatment } from '../../clinical/clinical-entities';
+import {
+	ExceptionType,
+	ProgramPropertyExceptionRecord,
+	EntityPropertyExceptionRecord,
+} from './types';
+import {
+	EntityExceptionRecord,
+	ProgramExceptionRecord,
+	isSpecimenExceptionRecord,
+	isTreatmentExceptionRecord,
+	isFollowupExceptionRecord,
+	Entity,
+} from '../property-exceptions/types';
+
+type entityKeyMap = {
+	specimen: 'specimenId';
+	treatment: 'treatmentId';
+	follow_up: 'followUpId';
+};
+
+const idKeys = {
+	specimen: 'submitter_specimen_id',
+	treatment: 'submitter_treatment_id',
+	follow_up: 'submitter_follow_up_id',
+};
+
+const entityKeys: entityKeyMap = {
+	specimen: 'specimenId',
+	treatment: 'treatmentId',
+	follow_up: 'followUpId',
+};
```

Comment:
these typings need to relate to the objects you're using.
there's a difference between 
```ts
type SpecimenEntityKey = 'specimenId'
```
and
```ts
type SpecimenEntityKey = Specimen['specimenId']
```